Source/Graphic/HCPath.c

```diff
@@ -230,21 +205,362 @@ HCPathElement HCPathElementAt(HCPathRef self, HCInteger elementIndex) {
+HCBoolean HCPathContainsPoint(HCPathRef self, HCPoint point) {
+    // Determine how many crossings there are for a ray from the point going in the +x direction
+    // Rather than construct a path and the intersection function, this function uses custom intersection code similar to the intersection function
+    HCInteger intersectionCount = 0;
+    HCPoint distantPoint = HCPointMake(self->bounds.size.width * 2.0, point.y);
+    HCPoint q0 = point;
+    HCPoint q1 = distantPoint;
+    HCInteger elementCount = HCPathElementCount(self);
+    HCInteger elementIndex = 0;
+    while (elementIndex < elementCount) {
+        // Determine if the current contour is open, and skip it if it is
+        HCInteger endIndex = 0;
+        HCBoolean closed = HCPathContourContainingElementIsClosed(self, elementIndex, NULL, &endIndex);
+        if (!closed) {
+            // Move to the next contour
+            elementIndex = endIndex;
+            continue;
+        }
+        
+        // Extract the initial point of the polyline and use it as the start point of the first line segment, or skip it if it is empty (e.g. a move-to element)
+        HCInteger polylinePointCount = HCPathPolylinePointCount(self, elementIndex);
+        if (polylinePointCount <= 0) {
+            // Move to the next element
+            elementIndex++;
+            continue;
+        }
+        HCPoint p0 = HCPathPolylinePointAt(self, elementIndex, 0);
+        
+        // Compare the line segments of the polyline against the ray
+        for (HCInteger pointIndex = 1; pointIndex < polylinePointCount; pointIndex++) {
+            // Determine the end point of the line segment from the polyline
+            HCPoint p1 = HCPathPolylinePointAt(self, elementIndex, pointIndex);
+            
+            // Find the intersection parameters for the lines formed by the line segment points
+            // TODO: Could performance be improved using a simpler intersection function that is for line-segment to ray?
+            HCReal t = 0.0;
+            HCReal u = 0.0;
+            HCContourLineLineIntersection(p0, p1, q0, q1, &t, &u);
+            
+            // Determine if they intersect within the bounds of the segments
+            HCBoolean segmentsIntersect = t >= 0.0 && t <= 1.0 && u >= 0.0 && u <= 1.0;
+            if (segmentsIntersect) {
+                intersectionCount++;
+            }
+            
+            // Use the end point of the polyline line segment as the start point of the next
+            p0 = p1;
+        }
+        
+        // Move to the next element
+        elementIndex++;
+    }
+    
+    // An odd intersection count indicates the point is within the path
+    return intersectionCount % 2 == 1;
+}
+
+HCBoolean HCPathContainsPointNonZero(HCPathRef self, HCPoint point) {\
+    (void)self; // unused
+    (void)point; // unused
```

Comment:
Do we need these?

Reply:
Yes, otherwise compiling with -Wall gives a warning.